app/wasm/test/custom_msg_test.go

```diff
@@ -514,6 +515,98 @@ func TestSwapMsg(t *testing.T) {
 	}
 }
 
+func TestJoinPoolMsg(t *testing.T) {
+	creator := RandomAccountAddress()
+	osmosis, ctx := SetupCustomApp(t, creator)
+
+	provider := RandomAccountAddress()
+
+	providerFunds := sdk.NewCoins(
+		sdk.NewInt64Coin("uatom", 333000000),
+		sdk.NewInt64Coin("uosmo", 555000000+3*poolFee),
+		sdk.NewInt64Coin("uregen", 777000000),
+		sdk.NewInt64Coin("ustar", 999000000),
+	)
+	fundAccount(t, ctx, osmosis, creator, providerFunds)
+	fundAccount(t, ctx, osmosis, provider, providerFunds)
+
+	// 20 star to 1 osmo
+	funds1 := []sdk.Coin{
+		sdk.NewInt64Coin("uosmo", 12000000),
+		sdk.NewInt64Coin("ustar", 240000000),
+	}
+	starPool := preparePool(t, ctx, osmosis, creator, funds1)
+
+	poolDenom := "gamm/pool/1"
+
+	creatorBalance := osmosis.BankKeeper.GetBalance(ctx, creator, poolDenom)
+
+	require.Equal(t, "100000000000000000000", creatorBalance.Amount.String())
+
+	poolInfo, err := osmosis.GAMMKeeper.GetPoolAndPoke(ctx, starPool)
+
+	require.NoError(t, err)
+
+	coinsInPool := poolInfo.GetTotalPoolLiquidity(ctx)
+
+	require.Equal(t, sdk.NewCoin("uosmo", sdk.NewInt(12000000)), coinsInPool[0])
+	require.Equal(t, sdk.NewCoin("ustar", sdk.NewInt(240000000)), coinsInPool[1])
+
+	totalSharesBefore := poolInfo.GetTotalShares()
+
+	require.Equal(t, "100000000000000000000", totalSharesBefore.String())
+
+	osmoStarLiquidity := sdk.NewCoins(sdk.NewInt64Coin("uosmo", 12_000), sdk.NewInt64Coin("ustar", 240_000))
+	reflect := instantiateReflectContract(t, ctx, osmosis, provider)
+
+	invalidMsg := wasmbindings.OsmosisMsg{JoinPoolNoSwap: &wasmbindings.JoinPoolNoSwap{
+		PoolId:         starPool,
+		ShareOutAmount: sdk.NewInt(100000000000000000),
+		TokenInMaxs:    sdk.NewCoins(sdk.NewCoin("random", sdk.NewInt(10))),
+	}}
+	expectedErr := executeCustom(t, ctx, osmosis, reflect, provider, invalidMsg, osmoStarLiquidity)
+	require.Error(t, expectedErr)
+	require.Equal(t, "dispatch: submessages: join pool: TokenInMaxs is less than the needed LP liquidity to this JoinPoolNoSwap, upperbound: 10random, needed 12000uosmo,240000ustar: calculated amount is larger than max amount", expectedErr.Error())
+
+	//ShareOutAmount = TotalShares * tokenInAmount / poolAsset.amount
+	//Either asset can be used for tokenInAmount and poolAsset.amount can be used to calculate this amount
+	//ShareOutAmount = 100000000000000000000 * 12000 / 12000000 = 100000000000000000000 * 240000 / 240000000
+
+	msg := wasmbindings.OsmosisMsg{JoinPoolNoSwap: &wasmbindings.JoinPoolNoSwap{
+		PoolId:         starPool,
+		ShareOutAmount: sdk.NewInt(100000000000000000),
+		TokenInMaxs:    osmoStarLiquidity,
+	}}
+	err = executeCustom(t, ctx, osmosis, reflect, provider, msg, sdk.NewCoins(sdk.NewCoin("random", sdk.NewInt(10))))
+
+	require.Error(t, err)
+	require.Equal(t, "0random is smaller than 10random: insufficient funds", err.Error())
+
+	err = executeCustom(t, ctx, osmosis, reflect, provider, msg, osmoStarLiquidity)
+
+	require.NoError(t, err)
+
+	//Simualate sending share tokens to user
```

Comment:
Just trying to understand - how does everything below help test the `JoinPoolNoSwap` message?

This seems like an integration level test if the goal is to ensure that components work well together

Reply:
Good point.  I can make the change to ensure the shareToken is sent to the `reflect` address, not the `provider` address.  I wanted to ensure that the share and liquidity is updated correctly.